# Part 1. Equation of a Slime

How many late days are you using for this assignment? 0

In [119]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn import svm
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
import numpy as np

## 1. Loading the dataset

In [120]:
# Using pandas load the dataset
s = pd.read_csv("science_data_large.csv")
df = pd.DataFrame(s)
# Output the first 15 rows of the data
s.head(15)

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


In [121]:
# Display a summary of the table information (number of datapoints, etc.)
s.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## 2. Splitting the dataset

In [122]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop("Size nm^3", axis = 1)
y = df["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
# For grading consistency use random_state=42 

## 3. Perform a Linear Regression

In [123]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame({'Temperature °C': [450], 'Mols KCL': [500]})
predicted_output = model.predict(sample)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
model.score(X, y)

0.8606369652963943

In [124]:
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_  
intercept = model.intercept_ 
print("coefficients:", coefficients)
print("intercept:", intercept)

coefficients: [ 866.14641337 1032.69506649]
intercept: -409391.47958340764


Write the linear equation of a slime: (example equation: $E = mc^2$) \
$-409391.47958 + 866.14641x_1 + 1032.69507x_2$

Report on score and explain meaning: \
The default score function for linear regression is $R^2$. The best possible $R^2$ score is 1, meaning the model predicts correctly 100% of the time. If the model is not taking features into account and predicting randomly, then the $R^2$ score would be 0. If the model's predictions perform worse than random, then the $R^2$ value would be negative. The linear regression model trained above received a score of ~0.86064, which is positive and closer to 1 than 0, which means the model performed well.

## 4. Use Cross Validation

In [125]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
cv_scores = cross_val_score(model, X, y, cv=5, scoring='r2')
print(cv_scores)
# Report on their finding and their significance

[0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]


Write findings here: \
Since I chose the same scoring as the previous model, we can fairly compare the two models. Based on the $R^2$ values, 2 out of 5 of the scores are more than the linear regression model, meaning that the model performed better than the linear regression model 40% of the time and slightly worse 60% of the time. This means that the two models perform about the same.

## 5. Using Polynomial Regression

In [126]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
model.fit(X_poly, y)
cv_scores = cross_val_score(model, X_poly, y, cv=5, scoring='r2')
print(cv_scores)

[1. 1. 1. 1. 1.]


In [127]:
# Report on the metrics and output the resultant equation as you did in Part 3.
pr_coefficients = model.coef_  
pr_intercept = model.intercept_ 
print("coefficients:", pr_coefficients)
print("intercept:", pr_intercept)
# feature_names = poly.get_feature_names_out(input_features=X.columns)
# print(feature_names)

coefficients: [ 0.00000000e+00  1.20000000e+01 -1.23114510e-07 -1.05666587e-11
  2.00000000e+00  2.85714287e-02]
intercept: 1.6574806068092585e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)\
$1.65748e^{-5} + 12x_1 -1.23114e^{-07}x_2 -1.05666e^{-11}x_1^2
  +2x_1x_2 + 0.028571x_2^2$

Report on the score and interpret:\
Since this model was validated using the same 5-fold cross validation outputting $R^2$ score, all 1's means that this model is 100% accurate, correctly predicting the size of slime 100% of the time. Therefore, we found the exact equation of the slime!

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [180]:
# Load the dataset. Then train and evaluate the classification models.
ckd = pd.read_csv("ckd_feature_subset.csv")
df = pd.DataFrame(ckd)
X = df.drop('Target_ckd', axis=1)
y = df['Target_ckd']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 1. Logistic Regression

In [129]:
lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)
lr_scores = cross_val_score(lr_model, X, y, cv=5, scoring='accuracy')
print(lr_scores)

[0.83870968 0.83870968 0.87096774 0.93333333 0.76666667]


## 2. Support Vector Machines

In [130]:
svm_model = svm.SVC()
svm_model.fit(X_train, y_train)
svm_scores = cross_val_score(svm_model, X, y, cv=5, scoring='accuracy')
print(svm_scores)

[0.90322581 1.         0.87096774 1.         0.86666667]


## 3. k-Nearest Neighbors

In [131]:
knn_model = KNeighborsClassifier()
knn_model.fit(X_train, y_train)
knn_scores = cross_val_score(knn_model, X, y, cv=5, scoring='accuracy')
print(knn_scores)

[0.90322581 0.96774194 0.90322581 0.96666667 0.9       ]


/Users/<redacted>/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
/Users/<redacted>/opt/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will n

## 4. Neural Networks

In [148]:
# Experiments with Neural Network.
nn_model_1 = MLPClassifier(max_iter=10000)
nn_model_1.fit(X_train, y_train)
nn_scores_1 = cross_val_score(nn_model_1, X, y, cv=5, scoring='accuracy')
print(nn_scores_1)

[0.96774194 0.93548387 0.90322581 0.96666667 0.93333333]


In [164]:
nn_model_2 = MLPClassifier(hidden_layer_sizes=(8, 4), max_iter=10000, random_state=42)
nn_model_2.fit(X_train, y_train)
nn_scores_2 = cross_val_score(nn_model_2, X, y, cv=5, scoring='accuracy')
print(nn_scores_2)

[1.         0.93548387 0.90322581 1.         0.93333333]


In [178]:
nn_model_3 = MLPClassifier(activation='identity', solver='lbfgs', learning_rate='adaptive', max_iter=10000, random_state=42)
nn_model_3.fit(X_train, y_train)
nn_scores_3 = cross_val_score(nn_model_3, X, y, cv=5, scoring='accuracy')
print(nn_scores_3)

[1.         0.93548387 0.96774194 0.96666667 0.93333333]


## Results and Conclusion for Classification Experiments

In [135]:
data = {
    'Array': ['Logistic Regression', 'Support Vector Machines', 'k-Nearest Neighbors'],
    'Mean': [np.mean(lr_scores), np.mean(svm_scores), np.mean(knn_scores)],
    'Standard Deviation': [np.std(lr_scores), np.std(svm_scores), np.std(knn_scores)]
}
results_table = pd.DataFrame(data)
print(results_table)

                     Array      Mean  Standard Deviation
0      Logistic Regression  0.849677            0.054005
1  Support Vector Machines  0.928172            0.059993
2      k-Nearest Neighbors  0.928172            0.031893


Conclusion:\
Logistic Regression performed the worst, while SVM and k-Nearest Neighbors performed surprisingly the exact same when you just look at the mean. However, since KNN has a smaller standard deviation, KNN overall performed the best. I believe this is the case because KNN models don't necessarily need a linear relationship between parameters as logistic regression does to perform better. Additionally, KNN models don't have as much of an impact with outliers as do logistic regression models do, which may explain why KNN has a lower standard deviation than the SVM model.

## Results and Conclusion for Neural Network Experiments

In [179]:
NN_data = {
    'Array': ['Model 1', 'Model 2', 'Model 3'],
    'Mean': [np.mean(nn_scores_1), np.mean(nn_scores_2), np.mean(nn_scores_3)],
    'Standard Deviation': [np.std(nn_scores_1), np.std(nn_scores_2), np.std(nn_scores_3)]
}
results_table = pd.DataFrame(NN_data)
print(results_table)

     Array      Mean  Standard Deviation
0  Model 1  0.941290            0.024040
1  Model 2  0.954409            0.038934
2  Model 3  0.960645            0.024554


Conclusion/Findings:\
For the first model, I used the default parameters with the MLPClassifier, only changing the max iterations as the default number was giving me an error. Then, for the second model, I played around with hidden layers and also entering 43 as the random state, which increased the mean by ~0.013 but also increased the standard deviation by ~0.014. The random state didn't change the results, however the hidden layers did. Then, for the last model, I changed the activation, solver, and learning rate from the defaults which increased the mean but then decreased the standard deviation from the second model. I noticed that changing the activation to identity slightly decreased the mean, but also slightly increased the standard deviation. Changing the solver to lbfgs did the most. With the default solver, the mean would be the lowest, but with it the mean increased by ~0.03 and the standard deviation decreased by ~0.005. Then, the learning rate didn't do much.